doc/modules/lmnn.rst

```diff
@@ -0,0 +1,161 @@
+and only a single parameter (``n_neighbors``) has to be selected by the user
+before training.
+
+Large Margin Nearest Neighbor classification has been proven to work well in
+practice for data sets of varying size and difficulty.
+
+
+.. Enter ../auto_examples/classification/images/sphx_glr_plot_lda_lmnn_001.png
+        :target: ../auto_examples/classification/plot_lda_lmnn.html
+        :scale: 80
+
+.. Enter centered:: |lmnn|
+
+.. The plot shows decision boundaries for nearest neighbor classification and
+    large margin nearest neighbor classification. The bottom row shows the
+    same comparison after first doing PCA on the inputs.
+
+.. Enter topic:: Examples:
+    .. Comment
+    :ref:`sphx_glr_auto_examples_classification_plot_lda_lmnn.py`: Comparison
+    of LDA and LMNN on synthetic data.
+
+Dimensionality reduction
+========================
+
+:class:`lmnn.LargeMarginNearestNeighbor` can be used to
+perform supervised dimensionality reduction. The input data are projected
+onto a linear subspace consisting of the directions which maximize the nearest
+neighbor classification accuracy. Therefore the separation between clusters
+of samples of the same class is also maximized.
+
+This is implemented in :func:`lmnn.LargeMarginNearestNeighbor.transform`.
+The desired dimensionality can be set using the ``n_features_out`` parameter
+in the constructor.
+
+.. Comment topic:: Examples:
+    ..
+    :ref:`sphx_glr_auto_examples_decomposition_plot_pca_vs_lda_vs_lmnn.py`:
+    Comparison of PCA, LDA and LMNN
+    for dimensionality reduction of the Olivetti faces dataset
+
+Mathematical formulation
+========================
+
+The LMNN objective function consists of two competing terms, the pull loss
+that pulls target neighbors closer to their reference samples and the push
+loss that pushes impostors away:
+
+.. math::
+
+    \varepsilon_{\text{pull}} (L) = \sum_{i, j \rightsquigarrow i} ||L(x_i -
+     x_j)||^2
+
+    \varepsilon_{\text{push}} (L) = \sum_{i, j \rightsquigarrow i}
+    \sum_{l} (1 - y_{il}) [1 + || L(x_i - x_j)||^2 - || L
+    (x_i - x_l)||^2]_+
+    \text{ where } y_{il} = 1 \text{ if } $y_i = y_l$ \text{ and 0 otherwise}.
+
+    \varepsilon(L) = (1 - \mu) \varepsilon_{\text{pull}} (L) +
+    \mu \varepsilon_{\text{push}} (L) \text{, } \quad \mu \in [0,1]
+
+
+The term inside the sum of the push loss is called the hingle loss and
+the notation amounts to :math:`[x]_+ = \max(0, x)`.
+The parameter :math:`\mu` is a trade-off between penalizing large distances
+to target neighbors and penalizing margin violations by impostors. In practice,
+the two terms are weighted equally.
+
+To use this model for classification, we just need to use a nearest neighbors
+classifier (:class:`classification.KNeighborsClassifier`) on the data
+transformed by the linear transformation :math:`L` found by LMNN.
+
+The loss function is not convex in :math:`L`, but there is an alternative
```

Comment:
Not sure if the convex version is worth discussing in details? it is not implemented and this may only confuse the reader. maybe a short note saying that the problem is nonconvex and that there exists a convex formulation (albeit constrained and more expensive to solve) is enough.